rustls/src/msgs/deframer.rs

```diff
@@ -96,6 +98,8 @@ impl MessageDeframer {
         // we do a zero length read.  That looks like an EOF to
         // the next layer up, which is fine.
         debug_assert!(self.used <= OpaqueMessage::MAX_WIRE_SIZE);
+        let need_capacity = Ord::min(OpaqueMessage::MAX_WIRE_SIZE, self.used + READ_SIZE);
+        self.buf.resize(need_capacity, 0);
```

Comment:
Hm, this adds (potential) allocator use and guaranteed-never-used memory zeroing (given `MessageDeframer::used` tracks meaningful memory in `buf`).

Just off some quick measurements, I think this halves bulk receive performance?

main:
```
$ BENCH_MULTIPLIER=30 ./target/release/examples/bench bulk TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
bulk	TLSv1_2	TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256	max_fragment_size:default	send	2643.45	MB/s
bulk	TLSv1_2	TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256	max_fragment_size:default	recv	3729.10	MB/s
```

this pr:
```
$ BENCH_MULTIPLIER=30 ./target/release/examples/bench bulk TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
bulk	TLSv1_2	TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256	max_fragment_size:default	send	2702.09	MB/s
bulk	TLSv1_2	TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256	max_fragment_size:default	recv	1958.14	MB/s
```

Reply:
Right, I think this just needs an if-statement to skip resizing if the buffer is large enough already, which should be a large part of the time in most cases?

Reply:
This has been improved, I get this on main:

```
djc-2021 main rustls $ BENCH_MULTIPLIER=30 ./target/release/examples/bench bulk TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
bulk    TLSv1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256   max_fragment_size:default     send     2611.72 MB/s
bulk    TLSv1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256   max_fragment_size:default     recv     2414.83 MB/s
```

and on this branch:

```
djc-2021 in-place-joining rustls $ BENCH_MULTIPLIER=30 ./target/release/examples/bench bulk TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
bulk    TLSv1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256   max_fragment_size:default     send     2598.67 MB/s
bulk    TLSv1_2 TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256   max_fragment_size:default     recv     2455.62 MB/s
```

That is, on my machine my branch can now receive 2% faster than on main. (It seems a little weird that on your machine, sends are about as fast as on mine, while receiving is ~50% faster.)